sda-commons-server-spring-data-mongo/docs/MIGRATION_GUIDE.md

```diff
@@ -0,0 +1,117 @@
+# Migration Guide
+
+## Getting started
+The spring-data-mongo package is available at the following coordinates:
+```groovy
+implementation "org.sdase.commons:sda-commons-server-spring-data-mongo"
+```
+
+Replacing the MorphiaBundle:
+
+```diff
+-  private final MorphiaBundle<AppConfiguration> morphiaBundle =
+-    MorphiaBundle.builder()
+-      .withConfigurationProvider(AppConfiguration::getMongo)
+-      .withEntityScanPackageClass(ConsentConfigurationEntity.class)
+-      .withValidation()
+-      .build();
+
++  private final SpringDataMongoBundle<AppConfiguration> mongoBundle = 
++    SpringDataMongoBundle.builder()
++      .withConfigurationProvider(AppConfiguration::getMongo)
++      .build();
+```
+
+Morphia's `Datastore` can be replaced by Spring Data Mongo's `MongoOperations`:
+```java
+public MongoOperations getMongoOperations() {
+  return this.mongoBundle.getMongoOperations();
+}
+```
+
+### Morphia compatibility
+When upgrading a service from 2.x.x as set of converters is provided to stay compatible with a
+database that was initialised with Morphia. To enable these converters, `.withMorphiaCompatibility()`
+can be used when building the SpringDataMongoBundle.
+
+```java
+SpringDataMongoBundle.builder()
+  .withConfigurationProvider(AppConfiguration::getSpringDataMongo)
+  .withMorphiaCompatibility()
+  build();
+```
+
+### Validation
+Automatic JSR validation is no longer provided in v3. If you still want to validate your models
+you can do so manually using the `ValidationFactory`:
+
+```java
+boolean isValid =
+  Validation.buildDefaultValidatorFactory()
+    .getValidator()
+    .validate(myEntity)
+    .isEmpty();
+```
+
+## Queries
+
+### Building queries with the datastore
+Simple operations can be realised by the MongoOperations directly.
+
+Saving an entity:
+```java
+public void save(SampleEntity entity) {
+    mongoOperations.save(entity);
```

Comment:
I would prefer `insert`. `save` is effectively an `upsert` and will also overwrite existing documents while `insert` is meant for creating new documents. As MongoDB has very little assertions in regards of transactions, you should avoid to overwrite full documents, because you may revert recent changes made in another request between the entity has been loaded and is then saved. The alternative would be to do an explicit update/upsert query for the fields that should be changed and don't even need to load the document before. This may not apply for ODS apps where it is ensured, that data is handled in sequence. Afaik Morphia does not have such a difference in the API.

Reply:
updated